cpp/oneapi/dal/backend/primitives/rng/device_engine_dpc.cpp

```diff
@@ -131,22 +193,36 @@ sycl::event shuffle(sycl::queue& queue,
                     Type* dst,
                     device_engine& engine_,
                     const event_vector& deps) {
-    Type idx[2];
-    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) ==
-        sycl::usm::alloc::device) {
+    // Check if destination pointer is on host, as device memory is not supported for dst here
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) == sycl::usm::alloc::host) {
         throw domain_error(dal::detail::error_messages::unsupported_data_type());
     }
-    void* state = engine_.get_host_engine_state();
+
     engine_.skip_ahead_gpu(count);
 
-    for (std::int64_t i = 0; i < count; ++i) {
-        uniform_dispatcher::uniform_by_cpu<Type>(2, idx, state, 0, count);
-        std::swap(dst[idx[0]], dst[idx[1]]);
-    }
-    auto event = queue.submit([&](sycl::handler& h) {
-        h.depends_on(deps);
+    // Array to store random indices for shuffling
+    auto idx_array = array<Type>::empty(queue, count);
+    Type* idx_ptr = idx_array.get_mutable_data();
+
+    // Generate random indices for shuffle
+    oneapi::mkl::rng::uniform<Type> distr(0, count);
```

Comment:
Same note about the `Type` as in `uniform_without_replacement` needs to be added here.